### PR TITLE
restructure and fix deterministic ordering or array.

### PR DESCRIPTION
**Pull request by puneet2019**
## 1. Overview

- restructure allowlistedvalidators.go ( is a file that edits db, only getters and setters of the specific struct should be present there)
- move getAllValidatorsState to delegation-strategy as it is only required there.
- add sorts to the specific functions that cause non-detereminism

### CODE REVIEW

**Review by Anmol1696 — APPROVED** [review]
LGTM

**Review by arhamchordia — APPROVED** [review]
LGTM
But PR was already present for these changes.

File: x/lscosmos/keeper/delegation_strategy.go [hunk 1]
```diff
@@ -14,14 +14,16 @@ import (
 // CONTRACT: allowlistedValList.len > 0, amount > 0
 func (k Keeper) DelegateMsgs(ctx sdk.Context, delegatorAddr string, amount sdk.Int, denom string) ([]sdk.Msg, error) {
 	// fetch a combined updated val set list and delegation state
-	updateValList, updatedDelegationState := k.GetAllValidatorsState(ctx)
+	updateValList, hostAccountDelegations := k.GetAllValidatorsState(ctx)
```
Comment (arhamchordia) [inline x/lscosmos/keeper/delegation_strategy.go:18]: Sort missing from here

File: x/lscosmos/keeper/delegation_strategy.go [hunk 2]
```diff
@@ -14,14 +14,16 @@ import (
...
-	delegationState.HostAccountDelegations = updatedDelegationState.HostAccountDelegations
+	delegationState.HostAccountDelegations = hostAccountDelegations
 
-	valAddressAmount, err := FetchValidatorsToDelegate(updateValList, delegationState, sdk.NewCoin(denom, amount))
+	updatedAllowListedValidators := types.AllowListedValidators{AllowListedValidators: updateValList}
+
+	valAddressAmount, err := FetchValidatorsToDelegate(updatedAllowListedValidators, delegationState, sdk.NewCoin(denom, amount))
 	if err != nil {
 		return nil, err
 	}
```
Comment (arhamchordia) [inline x/lscosmos/keeper/delegation_strategy.go:29]: Sort missinf from here.

File: x/lscosmos/keeper/delegation_strategy.go [hunk 3]
```diff
@@ -49,14 +51,16 @@ func (k Keeper) DelegateMsgs(ctx sdk.Context, delegatorAddr string, amount sdk.I
 // CONTRACT: allowlistedValList.len > 0, amount > 0
 func (k Keeper) UndelegateMsgs(ctx sdk.Context, delegatorAddr string, amount sdk.Int, denom string) ([]sdk.Msg, []types.UndelegationEntry, error) {
 	// fetch a combined updated val set list and delegation state
-	updateValList, updatedDelegationState := k.GetAllValidatorsState(ctx)
+	updateValList, hostAccountDelegations := k.GetAllValidatorsState(ctx)
```
Comment (arhamchordia) [inline x/lscosmos/keeper/delegation_strategy.go:55]: Same sort missing

File: x/lscosmos/keeper/delegation_strategy.go [hunk 4]
```diff
@@ -49,14 +51,16 @@ func (k Keeper) DelegateMsgs(ctx sdk.Context, delegatorAddr string, amount sdk.I
...
-	delegationState.HostAccountDelegations = updatedDelegationState.HostAccountDelegations
+	delegationState.HostAccountDelegations = hostAccountDelegations
+
+	updatedAllowListedValidators := types.AllowListedValidators{AllowListedValidators: updateValList}
 
-	valAddressAmount, err := FetchValidatorsToUndelegate(updateValList, delegationState, sdk.NewCoin(denom, amount))
+	valAddressAmount, err := FetchValidatorsToUndelegate(updatedAllowListedValidators, delegationState, sdk.NewCoin(denom, amount))
 	if err != nil {
 		return nil, nil, err
 	}
```
Comment (arhamchordia) [inline x/lscosmos/keeper/delegation_strategy.go:66]: Sort missing

File: x/lscosmos/keeper/delegation_strategy.go [hunk 5]
```diff
@@ -230,5 +232,77 @@ func DivideUndelegateAmountIntoValidatorSet(sortedValDiff types.WeightedAddressA
...
+	}
+
+	// Convert the updated delegation state map to slice of types.HostChainDelegation
+	var updatedDelegationState types.HostAccountDelegations
+	for key, value := range currentDelegationStateMap {
+		updatedDelegationState = append(updatedDelegationState, types.HostAccountDelegation{ValidatorAddress: key, Amount: value})
+	}
+
+	// sort both updatedValList and hostAccountDelegations
+	sort.Sort(updatedValSet)
```
Comment (arhamchordia) [inline x/lscosmos/keeper/delegation_strategy.go:303]: Remove sorting from here
Changed it to somewhere else in https://github.com/persistenceOne/pstake-native/pull/363

**Review by arhamchordia — CHANGES_REQUESTED** [review]

File: x/lscosmos/keeper/delegation_strategy.go (hunk 3 above)
Comment (puneet2019) [reply x/lscosmos/keeper/delegation_strategy.go:55]: added to .GetAllValidatorsState

File: x/lscosmos/keeper/delegation_strategy.go (hunk 1 above)
Comment (puneet2019) [reply x/lscosmos/keeper/delegation_strategy.go:18]: added to k.GetAllValidatorsState(ctx)

File: x/lscosmos/keeper/delegation_strategy.go (hunk 5 above)
Comment (puneet2019) [reply x/lscosmos/keeper/delegation_strategy.go:303]: yup,  added it back

File: x/lscosmos/keeper/delegation_strategy.go (hunk 4 above)
Comment (puneet2019) [reply x/lscosmos/keeper/delegation_strategy.go:66]: added  to FetchValidatorsToUndelegate

**Pull request merged**